indexer/indexer.py

```diff
@@ -0,0 +1,337 @@
+        This class should return the dictionary corresponding to the mapping
+        to be loaded into ElasticSearch. It should use the index_mapping_config
+        attribute as appropriate.
+
+        :return: Dictionary describing the ElaticSearch mapping.
+        """
+        raise NotImplementedError(
+            'users must define create_mapping to use this base class')
+
+    def load_mapping(self, *args, **kwargs):
+        """
+        Load the mapping into Elasticsearch.
+
+        This method is responsible for loading the mapping into Elasticsearch.
+        It first creates an index using the instance's attributes, and
+        then loads the mapping by also calling create_mapping(), which
+        will create the object describing the mapping.
+        """
+        # Creates the index
+        self.es_client.indices.create(index=self.index_name,
+                                      body=self.index_settings,
+                                      ignore=[400])
+        # Loads mappings for the index
+        self.es_client.put_mapping(index=self.index_name,
+                                   doc_type=self.doc_type,
+                                   body=self.create_mapping())
+
+
+class FileIndexer(Indexer):
+    """Create a file oriented index.
+
+    FileIndexer makes use of its index() function to perform indexing
+    of the data files that are presented to the class when creating an
+    instance.
+
+    End for now.
+
+    """
+
+    def index(self, bundle_uuid, bundle_version, **kwargs):
+        """Indexes the data files.
+
+        Triggers the actual indexing process
+
+        :param bundle_uuid: The bundle_uuid of the bundle that will be indexed.
+        :param bundle_version: The bundle of the version.
+        """
+        # Get the config driving indexing (e.g. the required entries)
+        req_entries = self.index_mapping_config['requested_entries']
+        # Iterate over each file
+        for _file in self.data_files.values():
+            # List the arguments for clarity
+            args = [req_entries, "", self.metadata_files]
+            # Get all the contents from the entries requested in the config
+            contents = {key: value for key, value in self.__get_item(*args)}
+            # Get the elasticsearch uuid for this particular data file
+            es_uuid = "{}:{}".format(bundle_uuid, _file['uuid'])
+            # Get the special fields added to the contents
+            special_ = self.special_fields(_file,
+                                           contents,
+                                           bundle_uuid=bundle_uuid,
+                                           bundle_version=bundle_version,
+                                           es_uuid=es_uuid)
+            contents = {**contents, **special_}
+            # Load the current file in question
+            # Ideally merge() should be called at this point
+            self.load_doc(doc_contents=contents, doc_uuid=es_uuid)
```

Comment:
Subclasses should not be responsible for writing the resulting doc. This method should return the resulting doc or docs and the baseclass should be in control of writing it.

See https://en.wikipedia.org/wiki/Template_method_pattern for details or come find me so we can talk in person.